client/state/data-layer/wpcom-http/README.md

```diff
@@ -0,0 +1,140 @@
+
+### Step C: HTTP middleware runs actual HTTP request
+
+The HTTP middleware understands what it means to run HTTP requests.
+It may take some liberties in how it decides to actually execute the requests.
+This liberty is what allows us to provide the global policies like retry behavior, offline behavior, and batching.
+
+In whatever means it decides fitting, the middleware issues the HTTP request.
+When either a response comes back in or the request fails, the middleware will dispatch a new follow-up action.
+This action comes from the original HTTP Request action but it will be wrapped with meta information including the response data or error message.
+That action gets reinjected into Redux and will presumably be handled by more middleware.
+
+In this case, we have assumed that a successful response from the server means that the like when through.
+This is a wrong assumption, but it is good enough for this discussion.
+As seen in the diagram, the API middleware instructed the HTTP middleware to fire a `LIKE_POST` action if the HTTP request succeeds and an `UNLIKE_POST` action if it fails.
+This `UNLIKE_POST` action will end up rolling back the first like, which in effect could not be completed because the server failed to update.
+
+
+### Step D: Reducers see action and update app state
+
+The actual `LIKE_POST` action (and possible the `UNLIKE_POST` as well) hit the reducers and the app state is updated.
+This will trigger a re-render on the post component which started the chain and we come full circle.
+
+## Implementation
+
+If you would like to take advantage of this system you will want to use the provided helper methods and structure your code into three pieces: a function which generates the descriptive HTTP requests; a function which handles successful responses; and a function which handles failing responses.
+
+Each of these functions will take the normal middleware arguments but the success and failure functions take an additional argument which is the response data or error respectively.
+
+```js
+// API Middleware, Post Like
+import { LIKE_POST } from 'state/action-types';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { QUEUE_REQUEST } from 'state/data-layer/wpcom-http/constants';
+
+/**
+ * @see https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/likes/new/ API description
+ */
+const likePost = ( { dispatch }, action, next ) => {
+	// dispatch intent to issue HTTP request
+	dispatch( http( {
+		method: 'POST',
+		path: '/sites/%s/posts/%d/likes/new',
+		pathArgs: [ action.siteId, action.postId ],
+		
+		// we can reuse the original action because
+		// we will use meta to indicate which part
+		// of the process we are in
+		onSuccess: action,
+		onFailure: action,
+		
+		// indicate what should happen if we have
+		// no network connection: queue to replay
+		// when the network reconnects
+		// (not implemented yet)
+		whenOffline: QUEUE_REQUEST,
```

Comment:
Drop seems reasonable. I think a simple timed-out queue might make more sense, especially for those _lie-fi_ conditions. this is a decision for later though probably (nothing is yet implemented)

> if a request comes in and we have no network connection then hold onto the request for up to five minutes and retry if it comes back online